PSMDB-1438-OIDC-8.0 (#997)
percona/psmdb-docs: docs/oidc.md

* **Authorization code**: a `mongo` client (for example, `mongosh` or Compass) opens a browser and redirects a user to the login portal of an external identity provider to pass authentication. This is the most common and secure flow for interactive user sessions. It is suitable for use cases when users have a web browser available on the machine where they are running the `mongo` client.

* **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a `mongo` client receives the URL of the login portal and the authentication code. The user authenticates on a separate device, following the URL and entering the authentication code. The example use case for such a workflow is when both a `mongo` client and Percona Server for MongoDB run in a cloud environment and the client needs to authenticate in Percona Server for MongoDB without managing long-term credentials like passwords. 

The following diagram illustrates the authentication flow.

![image](_images/OIDC-flow.png)

1. A user connects to Percona Server for MongoDB using a `mongo` client. The client must support OIDC.

> 2. The `mongo` client requests authentication from the IdP.

Reviewer: I'm not sure what level of details we would like to have here but when I read the flow I am missing one step between 1 and 2. The client receives the iformation about the IdP from the server (url, client_id, request scopes). In case multiple IdPs are configured on the server side the client must provide a username which will be used on the server side to pick appopriate IdP.

3. The IdP generates the authorization code. A user is redirected to the login portal of an external identity provider (IdP). Alternatively, a user is provided with a URL and the authentication code.
4. The user is requested to authenticate. For example, using two-factor authentication or by entering an authentication code.
5. A user is redirected back to the `mongo` client with a temporary single-use authorization code. 

> 6. The IdP verifies the authorization code, user's client ID and credentials.

Reviewer: IMHO it is a bit confusing that suddenly the IdP verifies the authorization code - this is because the mongo client is requesting the tokens by providing the authorization code.

Author: Updated

7. Upon success, the IdP returns the access and ID tokens to the `mongo` client.
8. The `mongo` client uses the access token to access Percona Server for MongoDB.

> The access and ID tokens are encoded as JSON Web Tokens (JWT). They contain information about user identities and authorization rights.

Reviewer: Well that is not always the case, this is actually what we support - the token must be JWT. If it is opaque token the authentication will fail.

Author: Changed to must be encoded

You can use the IdP infrastructure to authenticate and authorize users. In this case users are stored and managed on the IdP side. 

> Or you can bundle OIDC authentication with LDAP authorization. In this flow, users authenticate via an IdP and are authorized with the LDAP server.

Reviewer: Maybe we could indicate how to achive both ? The `useAuthorizationClaim` parameter decides how the user is authorized.

Reviewer 2: Agree. At best it can be a separate page to describe that and referenced on LDAP and OIDC pages.

Author: Added a separate tutorial

## Benefits

The use of OIDC and OAuth 2.0 provides the following benefits:

* streamlines authentication and authorization flow, 
* simplifies user management and configuration: everything is done in a single place
* enables you to use modern authentication techniques like 2FA, MFA and others supported by IdP
* improves security as credentials are not sent to nor stored in Percona Server for MongoDB. 
* reduces cross-application risk - access tokens are granted for specific resources using audience claims. If a token is compromised, the token has a limited lifetime and scope to limit access.

## Configuration 

To configure OIDC /OAuth 2.0 authentication and authorization, you must do the following:

1. Configure external identity provider
2. Configure authentication in Percona Server for MongoDB
3. Configure user roles and privileges in Percona Server for MongoDB that will be mapped to IdP groups

This section describes Percona Server for MongoDB configuration for OIDC authentication and the available configuration options. Refer to tutorials for detailed step-by-step instructions:

* [Configure OIDC with Okta](oidc-okta.md)
* [Configure OIDC with Microsoft Entra](oidc-entra.md)

> * [Configure OIDC with Ping Identity]

Reviewer: missing links

Author: Fixed

* [Configure OIDC with Keycloak]

### Percona Server for MongoDB configuration options

To configure OIDC authentication in Percona Server for MongoDB, specify the external identity provider configuration for the `oidcIdentityProviders` server parameter. You can set it only at startup. See the [Parameter tuning guide](set-parameter.md) for how to set server parameters.

> You can use several IdPs for OIDC authentication. In this case, you must add a configuration for every provider to the `oidcIdentityProviders` server parameter. You must also specify a match pattern for each provider. Usernames are then matched against the match pattern to identify which IdP to authenticate a user with. The order in which IdPs are listed defines their priority. The first IdP that matches the username is used for authentication.

Reviewer: The match pattern is required only if there are more than 1 IdPs which supports human flows.

Author: So for device auth this is not needed?

Author: Reworded

The `oidcIdentityProviders` server parameter contains an array of JSON objects with the following parameters:

| Parameter  | Required    | Description |
|------------|-------------|-------------| 
| `issuer`   | Yes         | The URL of the identity provider that the server should accept tokens from. It must be a valid URL that starts with `https://`. |

> | `audience` | Yes         | The audience claim for the identity provider. It is used to verify that the access token is intended for your application. You can specify only one value for the `audience` field. When you use multiple identity providers, the `audience` field must have a unique value. |

Reviewer: Audience must be unique only for IdPs with the same issuer.
For example the following is valid:
```
{
 issuer: "https://url1",
 audience: "some_audience"
},
{
 issuer: "https://url2",
 audience: "some_audience"
}
```

Author: Updated

| `authNamePrefix` | Yes   | The unique prefix for the authentication name. It is used to identify the identity provider in the authentication process. |

> | `useAuthorizationClaim`| No| If set to `true`, the server uses the claim in the access token to map user roles to MongoDB roles. If set to `false`, the server uses the `$external` database for authentication and authorization. The default value is `false`. |

Reviewer: The default value is `true`

Author: Fixed

| `authorizationClaim` | Yes when `useAuthorizationClaim` is `true` | The claim in the access token that contains the user roles or groups. It is used to map user roles to MongoDB roles.|

> | `clientId` | Yes | The client ID of the application registered with the identity provider. It is used to identify your application when requesting tokens. |

Reviewer: Required only inf `supportsHumanFlows=true`

Author: Updated

> | `matchPattern` | Yes (if more than one IdP is used) | A regular expression that matches usernames to identify which identity provider to use for authentication. |

Reviewer: if more than one IdP which supports human flows is used.

Author: Updated

The IdP is a centralized place to authenticate and authorize humans and applications to access multiple resources in your infrastructure. User credentials, access policies and roles are stored centralized on the IdP side. You can configure different access policies and tailor permissions for a group of users of a specific user. 

## Supported external identity providers

Percona Server for MongoDB supports the following external identity providers:

* [Okta :octicons-link-external-16:](https://www.okta.com/)
* [Microsoft Entra :octicons-link-external-16:](https://www.microsoft.com/en-gb/security/business/microsoft-entra)
* [Ping Identity:octicons-link-external-16:](https://www.pingidentity.com/en.html)
* [Keycloak :octicons-link-external-16:](https://www.keycloak.org/)

Any other external identity provider that supports OIDC and OAuth 2.0 may also work. However, we haven't tested them and cannot guarantee their compatibility.

## Authentication workflow

Percona Server for MongoDB supports two authentication workflows with OIDC:

> * **Authorization code**: a `mongo` client (for example, `mongosh` or Compass) opens a browser and redirects a user to the login portal of an external identity provider to pass authentication. This is the most common and secure flow for interactive user sessions. It is suitable for use cases when users have a web browser available on the machine where they are running the `mongo` client.

Reviewer 2: ```suggestion
* **Authorization code**: A MongoDB client (for example, MongoDB Shell or MongoDB Compass) opens a browser and redirects a user to the login portal of an external identity provider to pass authentication. This is the most common and secure flow for interactive user sessions. It is suitable for use cases when users have a web browser available on the machine where they are running their MongoDB client.
```

Author: We need to update all auth docs then. Changing one paragraph only introduces confusion

> * **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a `mongo` client receives the URL of the login portal and the authentication code. The user authenticates using a separate device, following the URL and entering the authentication code. The example use case for such a workflow is when both a `mongo` client works in an environment without a browser, e.g. on a CLI-only server. 

Reviewer 2: ```suggestion
* **Device authentication**: Instead of redirecting a user to authenticate on a login portal directly, a MongoDB client receives the URL of the login portal and the authentication code. The user authenticates using a separate device, following the URL and entering the authentication code. The example use case for such a workflow is when both a MongoDB client works in an environment without a browser, e.g., on a CLI-based server. 
```

> * **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a `mongo` client receives the URL of the login portal and the authentication code. The user authenticates on a separate device, following the URL and entering the authentication code. The example use case for such a workflow is when both a `mongo` client and Percona Server for MongoDB run in a cloud environment and the client needs to authenticate in Percona Server for MongoDB without managing long-term credentials like passwords. 

Reviewer 3: ```suggestion
* **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a MongoDB client receives the URL of the login portal and the authentication code. The user authenticates using a separate device, following the URL and entering the authentication code. Device authentication can be used when a MongoDB client works in an environment without a browser, e.g. on a CLI-only server. 
```

The following diagram illustrates the authentication flow.

![image](_images/OIDC-flow.png)

> 1. A user connects to Percona Server for MongoDB using a `mongo` client. The client must support OIDC.

Reviewer 3: From my perspective, the phrase "a `mongo` client" can be confusing. When we use `mongo` in a monospace font, we usually mean a specific program called "mongo". 
But in this doc, everything applies to each and every compatible client for MongoDB, and `mongo` is just one example of such a client. Others include but are  not limited to `mongosh`, Compass, DBeaver, etc. 

I suggest replacing "a `mongo` client" with "a MongoDB client" in the whole PR.

Author: We have a `mongo` client in other docs and tutorials. I'd stick to consistency here. We can improve the wording in a separate PR

> 2. Percona Server for MongoDB redirects the `mongo` client to the IdP for authentication and provides the IdP information such as URL, requests scopes, cleintID. 

Reviewer 2: ```suggestion
2. Percona Server for MongoDB redirects a MongoDB client to the IdP for authentication and provides the IdP information, such as URL, request scopes, and client ID. 
```

> 3. The `mongo` client requests authentication from the IdP.

Reviewer 2: ```suggestion
3. The MongoDB client requests authentication from the IdP.
```

4. The IdP generates the authorization code. A user is redirected to the login portal of an external identity provider (IdP). Alternatively, a user is provided with a URL and the authentication code.
5. The user is requested to authenticate. For example, using two-factor authentication or by entering an authentication code.

> 6. A user is redirected back to the `mongo` client with a temporary single-use authorization code. 

Reviewer 2: ```suggestion
6. A user is redirected back to the MongoDB client with a temporary single-use authorization code. 
```

> 7. The `mongo` client sends the authorization code to the IdP and requests tokens. The request includes the `clientId` and credentials of the `mongo` client.

Reviewer 2: ```suggestion
7. The MongoDB client sends the authorization code to the IdP and requests tokens. The request includes the `clientId` and credentials of the `mongo` client.
```

8. The IdP verifies the authorization code, user's client ID and credentials.

> 9. Upon success, the IdP returns the access and ID tokens to the `mongo` client.

Reviewer 2: ```suggestion
9. Upon success, the IdP returns the access and ID tokens to the MongoDB client.
```

> 10. The `mongo` client uses the access token to access Percona Server for MongoDB.

Reviewer 2: ```suggestion
10. The MongoDB client uses the access token to access Percona Server for MongoDB.
```

> The access and ID tokens are encoded as JSON Web Tokens (JWT). They contain information about user identities and authorization rights.

Reviewer 2: ```suggestion
The access and ID tokens should be encoded as JSON Web Tokens (JWT). They contain information about user identities and authorization rights.
```

# OIDC authentication

OpenID Connect (OIDC) is an identity authentication protocol built on top of the OAuth 2.0 framework. OIDC is designed to verify user identities and provide authentication, ensuring that users are who they claim to be. OAuth 2.0 is used for user authorization to access resources. 

> With the OIDC / OAuth 2.0 support in Percona Server for MongoDB, users can authenticate and authorize in your infrastructure without sharing their credentials. To make this happen, you enable a single sign-on (SSO) for Percona Server for MongoDB using an external identity provider (IdP). 

Reviewer 2: ```suggestion
With the OIDC / OAuth 2.0 support in Percona Server for MongoDB, users can authenticate and authorize your infrastructure without sharing their credentials in the MongoDB client. Single Sign-On (SSO) requires an external Identity Provider (IdP) for Percona Server for MongoDB.
```

Author: But it's not just about sharing creds in MongoDB client, They are also not saved in PSMDB, correct @ktrushin ?

Reviewer 3: Sorry for the delay in response.

> They are also not saved in PSMDB

Yes, that is correct, but the phrase "without sharing their credentials" looks  incomplete without specifying entities, with which sharing is not done.

Reviewer 3: Please consider the following alternative:
```suggestion
The OIDC / OAuth 2.0 support enables Percona Server for MongoDB to authenticate and authorize users via tokens issued by an identity provider (IdP). The latter serves as a centralized storage of user credentials, which are shared with neither Percona Server for MongoDB nor MongoDB clients.
```

Reviewer 2: Agree with Kostia, here. Please adjust. Especially - we need to fix "infrastructure" part - we don't use OIDC for infra but for DB

> The IdP is a centralized place to authenticate and authorize humans and applications to access multiple resources in your infrastructure. User credentials, access policies and roles are stored centralized on the IdP side. You can configure different access policies and tailor permissions for a group of users of a specific user. 

Reviewer 3: > roles are stored centralized on the IdP side.

That is not completely true. For OIDC to work, one has to set up either roles or users on MongoDB side as the [manual](https://www.mongodb.com/docs/manual/core/oidc/workforce/database-user-workforce/#steps) suggests.
Statement that credentials are not stored on MongoDB is still true.

Author: Updated to "managed"

Currently, Percona Server for MongoDB supports [Okta :octicons-link-external-16:](https://www.okta.com/) external identity provider. We plan to extend the list of supported external identity providers in future releases.

## Authentication workflow

Percona Server for MongoDB supports two authentication workflows with OIDC:

> * **Authorization code**: a `mongo` client (for example, `mongosh` or Compass) opens a browser and redirects a user to the login portal of an external identity provider to pass authentication. This is the default authentication workflow.

Reviewer 3: ```suggestion
* **Authorization code with Proof Key for Code Exchange (PKCE)**: a MongoDB client (for example, `mongosh` or Compass) opens a browser and redirects a user to the login portal of an external identity provider to pass authentication. This is the default authentication workflow.
```

> * **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a `mongo` client receives the URL of the login portal and the authentication code. The user follows the URL and enters the  authentication code. The example use case for such a workflow is when both a `mongo` client and Percona Server for MongoDB run in a cloud environment and the client needs to authenticate in Percona Server for MongoDB without managing long-term credentials like passwords. 

Reviewer 2: ```suggestion
* **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a MongoDB client receives the URL of the login portal and the authentication code. The user follows the URL and enters the authentication code. The example use case for such a workflow is when a MongoDB client and Percona Server for MongoDB run in a cloud environment, and the client needs to authenticate in Percona Server for MongoDB without managing long-term credentials like passwords. 
```

Reviewer 3: ```suggestion
* **Device authentication**: instead of redirecting a user to authenticate on a login portal directly, a MongoDB client receives the URL of the login portal and the authentication code. The user follows the URL and enters the  authentication code. The example use case for such a workflow is when both a MongoDB client runs in an environment without a browser being available (e.g. Docker container or cloud infrastructure).
```

 

The following diagram illustrates the authentication flow.

![image](_images/OIDC-flow.png)

1. A user connects to Percona Server for MongoDB using a `mongo` client. The client must support OIDC.
2. The `mongo` client requests authentication from the IdP.
3. The IdP generates the authorization code. A user is redirected to the login portal of an external identity provider (IdP).
4.	The user is requested to authenticate. For example, using two-factor authentication or by entering an authentication code.
5. A user is redirected back to the `mongo` client with single-use authorization code. 
6.	The IdP verifies the authorization code, user's client ID and credentials.
5.	Upon success, the IdP returns the access and ID tokens to the `mongo` client.

> 6.	The `mongo` client uses the access token to access Percona Server for MongoDB.

Reviewer 3: I suggest doing the following changes:
1. replace "`mongo` client" with "MongoDB client" because `mongo` is a name of a particular utility
2. In the point 2, replace `the IdP` with `an external identity provider (IdP)`. In point 3, do the opposite replacement: reduce `an external identity provider (IdP)` to just `the IdP`. That way, we introduce the acronym `IdP` at the time it is first used (i.e. in point 2)
3. Fix the point numbering. Now we have: `4, 5, 6, 5, 6`. It probably should be `4, 5, 6, 7, 8`.
4. Replace a tab character between the point number and its text; applies to all points starting from 4.

Author: Done

## Benefits

The use of OIDC and OAuth 2.0 provides the following benefits:

* streamlines authentication and authorization flow, 

> * simplifies user management and configuration: everything is done in a single place

Reviewer 3: I'd exclude this point, please see the comment about roles above.

Author: Done
